vulnerabilities/importers/ubuntu.py

```diff
@@ -21,36 +20,112 @@
 #  VulnerableCode is a free software code scanning tool from nexB Inc. and others.
 #  Visit https://github.com/nexB/vulnerablecode/ for support and download.
 
-from urllib.request import urlopen
 
-import bs4
+import asyncio
+import bz2
+import dataclasses
+from typing import Iterable
+from typing import List
+from typing import Mapping
+from typing import Set
+import xml.etree.ElementTree as ET
+
+
+from aiohttp import ClientSession
+from aiohttp.client_exceptions import ClientResponseError
+import requests
+
+
+from vulnerabilities.data_source import OvalDataSource, DataSourceConfiguration
+
+
+@dataclasses.dataclass
+class UbuntuConfiguration(DataSourceConfiguration):
+    releases: list
+    etags: dict
+
+
+class UbuntuDataSource(OvalDataSource):
+
+    CONFIG_CLASS = UbuntuConfiguration
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # we could avoid setting translations, and have it
+        # set by default in the OvalParser, but we don't yet know
+        # whether all OVAL providers use the same format
+        self.translations = {'less than': '<'}
+        self.pkg_manager_api = VersionAPI()
 
+    def _fetch(self):
+        base_url = 'https://people.canonical.com/~ubuntu-security/oval/'
+        file_name = 'com.ubuntu.{}.cve.oval.xml.bz2'
+        releases = self.config.releases
+        for release in releases:
+            file_url = base_url + file_name.format(release)
+            if not self.create_etag(file_url):
+                continue
+            resp = requests.get(file_url)
+            extracted = bz2.decompress(resp.content)
+            yield (
+                {'type': 'deb', 'namespace': 'ubuntu'},
+                ET.ElementTree(ET.fromstring(extracted.decode('utf-8')))
+            )
+        # In case every file is latest, _fetch won't yield anything(due to checking for new etags),
+        # this would return None to added_advisories
+        # which will cause error, hence this
+        # function return an empty list
+        return []
 
-UBUNTU_ROOT_URL = 'https://people.canonical.com/~ubuntu-security/cve/main.html'
+    def set_api(self, packages):
+        asyncio.run(self.pkg_manager_api.load_api(packages))
```

Comment:
I'm not sure `asnycio.run()` is the right choice here. The [documentation](https://docs.python.org/3.8/library/asyncio-task.html#asyncio.run) says: "This function always creates a new event loop and closes it at the end. It should be used as a main entry point for asyncio programs, and should ideally only be called once."
But I don't know much about Python asyncio.

Reply:
>It should be used as a main entry point for asyncio programs.

That's what it's precisely doing in this case. 

As for 
>Should ideally only be called once
 
They also mention, 
>This function cannot be called when another asyncio event loop is running in the same thread.

Which IMHO, is fine with the implementation, because `self.pkg_manager_api.load_api(packages)` doesn't lead to creation of other event loops i.e if you follow the function  `self.pkg_manager_api.load_api(packages)`  nowhere it creates an event loop (via for eg. `asyncio.run`).

I wouldn't worry about this.